File: sdk-ts/src/replay.ts
/** Portable fixture replayer (mirrors conformance/replay.py). Any implementation
 * proves wire compatibility by replaying conformance/fixtures/*.json. */

export interface FixtureResult { name: string; ok: boolean; fails: string[] }

function getPath(obj: unknown, dotted: string): [boolean, unknown] {
  let cur: unknown = obj;
  for (const part of dotted.split(".")) {
    if (Array.isArray(cur) && /^\d+$/.test(part)) {
      const i = Number(part);
      if (i >= cur.length) return [false, undefined];
      cur = cur[i];
    } else if (cur !== null && typeof cur === "object" && part in (cur as Record<string, unknown>)) {
      cur = (cur as Record<string, unknown>)[part];
    } else return [false, undefined];
  }
  return [true, cur];
}

function sub<T>(obj: T, vars: Record<string, unknown>): T {
  if (typeof obj === "string") {
    let s: string = obj;
    for (const [k, v] of Object.entries(vars)) s = s.split(`{${k}}`).join(String(v));
    return s as unknown as T;
  }
  if (Array.isArray(obj)) return obj.map((v) => sub(v, vars)) as T;
  if (obj !== null && typeof obj === "object") {
    return Object.fromEntries(Object.entries(obj).map(([k, v]) => [k, sub(v, vars)])) as T;
  }
  return obj;
}

const deepEqual = (a: unknown, b: unknown): boolean => JSON.stringify(a) === JSON.stringify(b);

export async function replayFile(baseUrl: string, spec: { version: string; fixtures: unknown[] }): Promise<FixtureResult[]> {
  if (spec.version !== "0.1") throw new Error(`unsupported fixture version ${spec.version}`);
  const vars: Record<string, unknown> = {};
  const saved: Record<string, unknown> = {};
  const out: FixtureResult[] = [];
  for (const fx of spec.fixtures as Array<Record<string, any>>) {
    const fails: string[] = [];
    try {
      const req = fx.request;
      const resp = await fetch(baseUrl + sub(req.path, vars), {
        method: req.method,
        headers: { "Content-Type": "application/json", ...sub(req.headers ?? {}, vars) },
        body: req.body === undefined ? undefined : JSON.stringify(sub(req.body, vars)),
      });
      const body: unknown = await resp.json().catch(() => ({}));
      const headers: Record<string, string | null> = {};
      resp.headers.forEach((v, k) => { headers[k] = v; });
      const exp = fx.expect ?? {};
      if (exp.status !== undefined && resp.status !== exp.status) {
        fails.push(`status ${resp.status} != ${exp.status} (body=${JSON.stringify(body).slice(0, 200)})`);
      }
      for (const key of exp.has ?? []) {
        if (body === null || typeof body !== "object" || !(key in (body as Record<string, unknown>))) {
          fails.push(`missing key '${key}'`);
        }
      }
      for (const path of exp.absent ?? []) {
        if (getPath(body, path)[0]) fails.push(`forbidden path present: '${path}'`);
      }
      for (const [path, wantRaw] of Object.entries(exp.where ?? {})) {
        const want = sub(wantRaw, vars);
        const [ok, got] = getPath(body, path);
        if (!ok) fails.push(`missing path '${path}'`);
        else if (!deepEqual(got, want)) fails.push(`${path}: ${JSON.stringify(got)} != ${JSON.stringify(want)}`);
      }
      for (const [hk, hv] of Object.entries(exp.headers ?? {})) {
        if ((headers[hk.toLowerCase()] ?? null) !== hv) fails.push(`header ${hk} != ${hv}`);
      }
      if (exp.identical_to !== undefined && !deepEqual(body, saved[exp.identical_to])) {
        fails.push(`body differs from saved '${exp.identical_to}'`);
      }
      for (const [v, path] of Object.entries(fx.capture ?? {})) {
        const [ok, got] = getPath(body, path as string);
        if (ok) vars[v] = got; else fails.push(`capture failed: '${path}'`);
      }
      if (fx.save_as) saved[fx.save_as] = body;
    } catch (e) {
      fails.push(`transport: ${(e as Error).message}`);
    }
    out.push({ name: fx.name, ok: fails.length === 0, fails });
  }
  return out;
}

File: conformance/replay.py
"""Portable conformance replayer: boots the Python reference in-process
and replays conformance/fixtures/*.json. Ports MUST reimplement this
(~80 lines) against the same fixture files.

Run:  python conformance/replay.py [fixture-file ...]  (default: all)
"""

import json
import sys
import threading
import urllib.request
import urllib.error

sys.path.insert(0, ".")
from reference.agent import ThreadingHTTPServer, Handler  # noqa: E402

PORT = 8487
BASE = f"http://127.0.0.1:{PORT}"


def get_path(obj, dotted):
    cur = obj
    for part in dotted.split("."):
        if not isinstance(cur, dict) or part not in cur:
            return (False, None)
        cur = cur[part]
    return (True, cur)


def sub(obj, variables):
    if isinstance(obj, str):
        for k, v in variables.items():
            obj = obj.replace("{" + k + "}", str(v))
        return obj
    if isinstance(obj, dict):
        return {k: sub(v, variables) for k, v in obj.items()}
    if isinstance(obj, list):
        return [sub(v, variables) for v in obj]
    return obj


def run_fixture(fx, variables, saved):
    req = fx["request"]
    data = json.dumps(sub(req.get("body"), variables)).encode() if "body" in req else None
    r = urllib.request.Request(
        BASE + sub(req["path"], variables), data=data,
        headers={"Content-Type": "application/json",
                 **sub(req.get("headers", {}), variables)},
        method=req["method"])
    try:
        with urllib.request.urlopen(r) as resp:
            status, body, headers = resp.status, json.loads(resp.read() or b"{}"), dict(resp.headers)
    except urllib.error.HTTPError as e:
        status, body, headers = e.code, json.loads(e.read() or b"{}"), dict(e.headers)
    exp = fx.get("expect", {})
    fails = []
    if "status" in exp and status != exp["status"]:
        fails.append(f"status {status} != {exp['status']} (body={str(body)[:200]})")
    for key in exp.get("has", []):
        if key not in body:
            fails.append(f"missing key {key!r}")
    for path, want in exp.get("where", {}).items():
        want = sub(want, variables)
        ok, got = get_path(body, path)
        if not ok:
            fails.append(f"missing path {path!r}")
        elif got != want:
            fails.append(f"{path}: {got!r} != {want!r}")
    for hk, hv in exp.get("headers", {}).items():
        if headers.get(hk) != hv:
            fails.append(f"header {hk}: {headers.get(hk)!r} != {hv!r}")
    if "identical_to" in exp and body != saved.get(exp["identical_to"]):
        fails.append(f"body differs from saved {exp['identical_to']!r}")
    for var, path in fx.get("capture", {}).items():
        ok, got = get_path(body, path)
        if ok:
            variables[var] = got
        else:
            fails.append(f"capture failed: {path!r}")
    if "save_as" in fx:
        saved[fx["save_as"]] = body
    return fails


def main():
    import glob
    files = sys.argv[1:] or sorted(glob.glob("conformance/fixtures/*.json"))
    srv = ThreadingHTTPServer(("127.0.0.1", PORT), Handler)
    threading.Thread(target=srv.serve_forever, daemon=True).start()
    total_pass, total_fail = 0, 0
    for path in files:
        spec = json.load(open(path))
        assert spec.get("version") == "0.1", f"{path}: unsupported fixture version"
        variables, saved = {}, {}
        print(f"[{path}]")
        for fx in spec["fixtures"]:
            fails = run_fixture(fx, variables, saved)
            if fails:
                total_fail += 1
                print(f"  FAIL {fx['name']}: {'; '.join(fails)}")
            else:
                total_pass += 1
                print(f"  PASS {fx['name']}")
    srv.shutdown()
    print(f"\n{total_pass} passed, {total_fail} failed")
    sys.exit(1 if total_fail else 0)


if __name__ == "__main__":
    main()
